File: src/Solutions/Euler180.hs
module Solutions.Euler180 where
import Data.Ratio
import Utils.List(uniques)

-- f_n(x, y, z) = (x^n + y^n - z^n)(x + y + z)
-- By Fermat's Last Theorem, solutions exist only if n \in \{-2, -1, 1, 2}.

bound = 35 :: Integer
ans :: Rational
ans = sum $ uniques
    [ x + y + z
    | xb <- [2..bound],
      xa <- [1..xb - 1],
      gcd xa xb == 1,
      let x = xa % xb,
      yb <- [2..bound],
      ya <- [1..yb - 1],
      gcd ya yb == 1,
      let y = ya % yb,
      x <= y,
      zb <- [2..bound],
      za <- [1..zb - 1],
      gcd za zb == 1,
      let z = za % zb,
      x^2 + y^2 == z^2 || x + y == z || 1 / x + 1 / y == 1 / z || 1 / x^2 + 1 / y^2 == 1 / z^2]

euler180 :: IO String
euler180 = return $ show $ numerator ans + denominator ans

File: rust/src/solutions/euler180.rs
use num::rational::Ratio;

// f_n(x, y, z) = (x^n + y^n - z^n)(x + y + z)
// By Fermat's Last Theorem, solutions exist only if n \in \{-2, -1, 1, 2}.

pub fn euler180() -> String {
    let mut vec = Vec::<Ratio<u64>>::new();
    let fractions = (2u64..36)
        .flat_map(|b| (1..b).map(move |a| (a, b)))
        .filter(|(a, b)| num::Integer::gcd(a, b) == 1)
        .map(|(a, b)| Ratio::new_raw(a, b))
        .collect::<Vec<Ratio<u64>>>();
    let _1 = Ratio::new_raw(1, 1);
    for &x in &fractions {
        for &y in &fractions {
            if x <= y {
                for &z in &fractions {
                    if x * x + y * y == z * z
                        || x + y == z
                        || _1 / x + _1 / y == _1 / z
                        || _1 / (x * x) + _1 / (y * y) == _1 / (z * z)
                    {
                        vec.push(x + y + z);
                    }
                }
            }
        }
    }
    vec.sort();
    vec.dedup();
    let sum = vec.into_iter().sum::<Ratio<u64>>();
    (sum.numer() + sum.denom()).to_string()
}
